segregation/inference/randomization.py

```diff
@@ -0,0 +1,382 @@
+        for each group of interest. Note that if not passing a `total` argument,
+        groups are assumed to be exhaustive. If total is not set and groups are not
+        exhaustive, the function will estimate incorrect probabilities of choosing
+        each geographic unit.
+
+    Returns
+    -------
+    geopandas.GeoDataFrame
+        geodataframe with systematically randomized population groups
+
+    Notes
+    -------
+    Simulates the random allocation of each group across geographic units, given the total population
+    of each group (randomizes location totals for each group). Given the total population of
+    each group in the region, take draws from a multinomial distribution where the probability of
+    choosing each geographic unit is equal to the total regional share currently residing in the unit.
+    Results are guaranteed to respect regional group totals, but will include variation in the total
+    population of each geographic unit.
+
+    For more, see Allen, R., Burgess, S., Davidson, R., & Windmeijer, F. (2015). More reliable inference for the dissimilarity index of segregation. The Econometrics Journal, 18(1), 40–66. https://doi.org/10.1111/ectj.12039
+
+    Reference: :cite:`allen2015more`
+    """
+    if groups:
+        if not total:
+            warn(
+                "No `total` argument passed. Assuming population groups are exhaustive"
+            )
+            total = "total"
+        df[total] = df[groups].sum(axis=1)
+    if group:
+        assert (
+            total
+        ), "If simulating a single group, you must also supply a total population column"
+        df["other_group_pop"] = df[total] - df[group]
+        groups = [group, "other_group_pop"]
+
+    p_j = df[total] / df[total].sum()
+    data_aux = {}
+    for group in groups:
+        n = df[group].sum()
+        sim = np.random.multinomial(n, p_j)
+        data_aux[group] = sim.tolist()
+    df_aux = pd.DataFrame.from_dict(data_aux)
+    df_aux[total] = df_aux[groups].sum(axis=1)
+    if isinstance(df, gpd.GeoDataFrame):
+        df_aux = df[[df.geometry.name]].reset_index().join(df_aux)
+
+    return df_aux
+
+
+def simulate_bootstrap_resample(df, **kwargs):
+    """Generate bootstrap replications of the units with replacement of the same size of the original data.
+
+    Parameters
+    ----------
+    df : geopandas.GeoDataFrame or pandas.DataFrame
+        (geo)dataframe with population counts to be randomized
+
+    Returns
+    -------
+    DataFrame
+        DataFrame with bootstrap resampled observations
+
+    Notes
+    -------
+    Simulate a synthetic dataset by drawing from rows of the input data with replacement
+    until reaching the same number of observations in the original dataframe. Note that if
+    input is a geodataframe, then the output will not be planar-enforced, as more than one of
+    the same unit may appear in the sample.
+    """
+    df = df.copy()
```

Comment:
What's the reason for creating an identically named data variable? (other instances also)